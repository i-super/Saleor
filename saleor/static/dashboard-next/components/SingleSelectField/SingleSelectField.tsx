import FormControl from "@material-ui/core/FormControl";
import FormHelperText from "@material-ui/core/FormHelperText";
import InputLabel from "@material-ui/core/InputLabel";
import MenuItem from "@material-ui/core/MenuItem";
import Select, { SelectProps } from "@material-ui/core/Select";
import { withStyles } from "@material-ui/core/styles";
import * as React from "react";

const decorate = withStyles({
  formControl: {
    width: "100%"
  }
});

interface SingleSelectFieldProps {
  choices: Array<{
    value: string;
    label: string;
  }>;
  disabled?: boolean;
  error?: boolean;
  hint?: string;
  label?: string;
  name?: string;
  selectProps?: SelectProps;
  value?: string;
  onChange(event: any);
}

export const SingleSelectField = decorate<SingleSelectFieldProps>(
  ({
    classes,
    disabled,
    error,
    label,
    choices,
    value,
    onChange,
    name,
    hint,
    selectProps
  }) => {
    const choicesByKey: { [key: string]: string } =
      choices === undefined
        ? {}
        : choices.reduce((prev, curr) => {
            prev[curr.value] = curr.label;
            return prev;
          }, {});

    return (
      <FormControl
        className={classes.formControl}
        error={error}
        disabled={disabled}
      >
        <InputLabel shrink={!!value}>{label}</InputLabel>
        <Select
          fullWidth
          renderValue={choiceValue =>
            choiceValue ? choicesByKey[choiceValue.toString()] : ""
          }
          value={value || ""}
          name={name}
          onChange={onChange}
          {...selectProps}
        >
          {choices.map(choice => (
            <MenuItem value={choice.value} key={choice.value}>
              {choice.label}
            </MenuItem>
          ))}
        </Select>
        {hint && <FormHelperText>{hint}</FormHelperText>}
      </FormControl>
    );
  }
);

export default SingleSelectField;
