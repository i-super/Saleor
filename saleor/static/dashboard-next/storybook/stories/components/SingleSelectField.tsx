import { storiesOf } from "@storybook/react";
import * as React from "react";

import SingleSelectField from "../../../components/SingleSelectField";

const choices = [
  { value: "1", label: "Apparel" },
  { value: "2", label: "Groceries" },
  { value: "3", label: "Books" },
  { value: "4", label: "Accessories" }
];

storiesOf("Generics / SingleSelectField", module)
  .add("with no value", () => (
    <SingleSelectField choices={choices} onChange={() => {}} />
  ))
  .add("with value", () => (
    <SingleSelectField
      choices={choices}
      onChange={() => {}}
      value={choices[0].value}
    />
  ))
  .add("with label", () => (
    <SingleSelectField
      choices={choices}
      onChange={() => {}}
      label="Lorem ipsum"
    />
  ))
  .add("with hint", () => (
    <SingleSelectField
      choices={choices}
      onChange={() => {}}
      hint="Lorem ipsum"
    />
  ))
  .add("with label and hint", () => (
    <SingleSelectField
      choices={choices}
      onChange={() => {}}
      label="Lorem"
      hint="Ipsum"
    />
  ))
  .add("with value, label and hint", () => (
    <SingleSelectField
      choices={choices}
      onChange={() => {}}
      value={choices[0].value}
      label="Lorem"
      hint="Ipsum"
    />
  ))
  .add("with error hint", () => (
    <SingleSelectField
      choices={choices}
      onChange={() => {}}
      hint="Lorem error"
      error={true}
    />
  ));
