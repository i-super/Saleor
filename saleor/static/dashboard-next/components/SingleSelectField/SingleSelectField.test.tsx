import * as React from "react";
import * as renderer from "react-test-renderer";

import SingleSelectField from "./";

const choices = [
  { value: "1", label: "Apparel" },
  { value: "2", label: "Groceries" },
  { value: "3", label: "Books" },
  { value: "4", label: "Accessories" }
];

describe("<MultiSelectField />", () => {
  it("renders with no value", () => {
    const component = renderer.create(
      <SingleSelectField choices={choices} onChange={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with value", () => {
    const component = renderer.create(
      <SingleSelectField
        choices={choices}
        onChange={() => {}}
        value={choices[0].value}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with label", () => {
    const component = renderer.create(
      <SingleSelectField
        choices={choices}
        onChange={() => {}}
        label="Lorem ipsum"
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with hint", () => {
    const component = renderer.create(
      <SingleSelectField
        choices={choices}
        onChange={() => {}}
        hint="Lorem ipsum"
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with label and hint", () => {
    const component = renderer.create(
      <SingleSelectField
        choices={choices}
        onChange={() => {}}
        label="Lorem"
        hint="Ipsum"
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with value, label and hint", () => {
    const component = renderer.create(
      <SingleSelectField
        choices={choices}
        onChange={() => {}}
        value={choices[0].value}
        label="Lorem"
        hint="Ipsum"
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with error hint", () => {
    const component = renderer.create(
      <SingleSelectField
        choices={choices}
        onChange={() => {}}
        hint="Lorem error"
        error={true}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
