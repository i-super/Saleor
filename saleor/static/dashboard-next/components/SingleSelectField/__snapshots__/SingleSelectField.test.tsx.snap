// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelectField /> renders with error hint 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiFormLabel-error-15 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10"
    data-shrink={false}
  />
  <div
    className="MuiInput-root-23 MuiInput-error-28 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203",
            }
          }
        />
      </div>
      <input
        aria-invalid={true}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-47 MuiFormHelperText-error-48"
  >
    Lorem error
  </p>
</div>
`;

exports[`<MultiSelectField /> renders with hint 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10"
    data-shrink={false}
  />
  <div
    className="MuiInput-root-23 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203",
            }
          }
        />
      </div>
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-47"
  >
    Lorem ipsum
  </p>
</div>
`;

exports[`<MultiSelectField /> renders with label 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10"
    data-shrink={false}
  >
    Lorem ipsum
  </label>
  <div
    className="MuiInput-root-23 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203",
            }
          }
        />
      </div>
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`<MultiSelectField /> renders with label and hint 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10"
    data-shrink={false}
  >
    Lorem
  </label>
  <div
    className="MuiInput-root-23 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203",
            }
          }
        />
      </div>
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-47"
  >
    Ipsum
  </p>
</div>
`;

exports[`<MultiSelectField /> renders with no value 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10"
    data-shrink={false}
  />
  <div
    className="MuiInput-root-23 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203",
            }
          }
        />
      </div>
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`<MultiSelectField /> renders with value 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10 MuiInputLabel-shrink-9"
    data-shrink={true}
  />
  <div
    className="MuiInput-root-23 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Apparel
      </div>
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value="1"
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`<MultiSelectField /> renders with value, label and hint 1`] = `
<div
  className="MuiFormControl-root-2 Component-formControl-1"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <label
    className="MuiFormLabel-root-12 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10 MuiInputLabel-shrink-9"
    data-shrink={true}
  >
    Lorem
  </label>
  <div
    className="MuiInput-root-23 MuiInput-fullWidth-30 MuiInput-formControl-24 MuiInput-underline-27"
  >
    <div
      className="MuiSelect-root-18"
    >
      <div
        aria-haspopup="true"
        aria-owns={null}
        aria-pressed="false"
        className="MuiSelect-select-19 MuiSelect-selectMenu-20 MuiInput-input-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Apparel
      </div>
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete={undefined}
        autoFocus={undefined}
        defaultValue={undefined}
        id={undefined}
        name={undefined}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="hidden"
        value="1"
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37 MuiSelect-icon-22"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-47"
  >
    Ipsum
  </p>
</div>
`;
